### PR TITLE
Add watcher service

### PR DESCRIPTION
**Pull request by rofafor**
One can use this e.g. for tracking configuration file modifications

### CODE REVIEW

File: v2/service/module/watcher/watcher.go [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+var (
+	ErrMissingWithFilename = fmt.Errorf("watcher.Watcher missing WithFilename option")
+	ErrMissingWithFunc     = fmt.Errorf("watcher.Watcher missing WithFunc option")
+)
+
+type Watcher struct {
+	w        *fsnotify.Watcher
+	stopped  chan struct{}
+	filename *string
```
Comment (heppu) [inline v2/service/module/watcher/watcher.go:18]: This could be just `string` instead of `*string` since empty string is also invalid value.

File: v2/service/module/watcher/watcher.go [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+	w.stopped = make(chan struct{})
+
+	for _, opt := range w.opts {
+		if err := opt(w); err != nil {
+			return fmt.Errorf("watcher.Watcher Option error: %w", err)
+		}
+	}
+
+	switch {
+	case w.filename == nil:
```
Comment (heppu) [inline v2/service/module/watcher/watcher.go:44]: With change above this would be just `w.filename == ""` and then we would also catch `WithFilename("")` case.

**Review by heppu — CHANGES_REQUESTED** [review]

File: v2/service/module/watcher/watcher.go [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+			if event.Op&fsnotify.Write == fsnotify.Write {
+				err := w.fn()
+				if err != nil {
+					return err
+				}
+			}
+		case err := <-w.w.Errors:
+			return fmt.Errorf("watcher.Watcher error: %w", err)
+		case <-w.stopped:
+			return nil
```
Comment (heppu) [inline v2/service/module/watcher/watcher.go:66]: Based on the documentation on https://pkg.go.dev/github.com/fsnotify/fsnotify#Watcher.Close using `w.stopped` to signal closing isn't required. Instead we should do:
```go
select {
case event, ok := <-w.w.Events:
	if !ok {
		return nil
	}
```

**Review by heppu — CHANGES_REQUESTED** [review]

File: v2/service/module/watcher/watcher_test.go [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+		watcher     *watcher.Watcher
+		expectedErr error
+	}{
+		{
+			name:        "ErrOpt",
+			watcher:     watcher.New(func(t *watcher.Watcher) error { return errOpt }),
+			expectedErr: errOpt,
+		},
+		{
+			name:        "ErrMissingWithFunc",
```
Comment (heppu) [inline v2/service/module/watcher/watcher_test.go:78]: Add also test case with empty string.

**Review by heppu — APPROVED** [review]
